File: clients/ios/StellarChat/StellarChat/Models/PNRelayClient.swift
import CryptoKit
import Foundation

/// Client for the privacy-preserving push notification relay.
/// All registrations use ephemeral X25519 keys — the PN relay never learns our Nostr identity.
actor PNRelayClient {
    private let relayURL: URL
    private var relayPublicKey: Curve25519.KeyAgreement.PublicKey?
    private let session = URLSession.shared

    init(relayURL: URL) {
        self.relayURL = relayURL
    }

    /// Fetch the PN relay's X25519 public key from GET /v1/info.
    func fetchInfo() async throws -> Curve25519.KeyAgreement.PublicKey {
        let url = relayURL.appendingPathComponent("v1/info")
        let (data, _) = try await session.data(from: url)
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let pubkeyB64 = json?["x25519_public_key"] as? String,
              let pubkeyData = Data(base64Encoded: pubkeyB64) else {
            throw PNRelayError.invalidInfoResponse
        }
        let key = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: pubkeyData)
        relayPublicKey = key
        return key
    }

    /// Register a push subscription for a single group topic.
    func subscribe(
        subscriptionID: String,
        filter: [String: Any],
        pushToken: String,
        notificationKey: Data,
        platform: String = "ios"
    ) async throws {
        let relayPubkey = try await getRelayPublicKey()

        let payload: [String: Any] = [
            "action": "subscribe",
            "subscription_id": subscriptionID,
            "filter": filter,
            "token": pushToken,
            "notification_key": notificationKey.base64EncodedString(),
            "platform": platform
        ]

        try await sendEncryptedEnvelope(payload: payload, relayPubkey: relayPubkey)
    }

    /// Update an existing subscription's filter (e.g., after topic rotation).
    func update(
        subscriptionID: String,
        filter: [String: Any]
    ) async throws {
        let relayPubkey = try await getRelayPublicKey()

        let payload: [String: Any] = [
            "action": "update",
            "subscription_id": subscriptionID,
            "filter": filter
        ]

        try await sendEncryptedEnvelope(payload: payload, relayPubkey: relayPubkey)
    }

    /// Unsubscribe from push notifications.
    func unsubscribe(subscriptionID: String) async throws {
        let relayPubkey = try await getRelayPublicKey()

        let payload: [String: Any] = [
            "action": "unsubscribe",
            "subscription_id": subscriptionID
        ]

        try await sendEncryptedEnvelope(payload: payload, relayPubkey: relayPubkey)
    }

    // MARK: - Private

    private func getRelayPublicKey() async throws -> Curve25519.KeyAgreement.PublicKey {
        if let key = relayPublicKey { return key }
        return try await fetchInfo()
    }

    /// Encrypt a registration payload using ephemeral X25519 + AES-256-GCM and POST it.
    private func sendEncryptedEnvelope(
        payload: [String: Any],
        relayPubkey: Curve25519.KeyAgreement.PublicKey
    ) async throws {
        let payloadData = try JSONSerialization.data(withJSONObject: payload)
        let envelope = try GroupCrypto.encryptInvitation(
            payloadData,
            recipientKeyAgreementPubkey: Data(relayPubkey.rawRepresentation),
            senderSigningKey: nil  // No signing — anonymous registration
        )
        let envelopeJSON = try JSONEncoder().encode(envelope)

        var request = URLRequest(url: relayURL.appendingPathComponent("v1/subscription"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = envelopeJSON
        request.timeoutInterval = 15

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw PNRelayError.serverError(body)
        }
    }
}

enum PNRelayError: LocalizedError {
    case invalidInfoResponse
    case serverError(String)
    case notConfigured

    var errorDescription: String? {
        switch self {
        case .invalidInfoResponse: return "Invalid PN relay info response"
        case .serverError(let msg): return "PN relay error: \(msg)"
        case .notConfigured: return "PN relay URL not configured"
        }
    }
}

File: clients/android/StellarChat/app/src/main/java/com/stellarmls/chat/push/PNRelayClient.kt
package com.stellarmls.chat.push

import com.stellarmls.chat.crypto.GroupCrypto
import com.stellarmls.chat.crypto.KeyManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class PNRelayClient(private val relayURL: String) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private var relayPublicKey: ByteArray? = null

    /// Fetch PN relay's X25519 public key.
    suspend fun fetchInfo(): ByteArray = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$relayURL/v1/info")
            .get()
            .build()
        val response = client.newCall(request).execute()
        val body = response.body?.string() ?: throw Exception("Empty info response")
        val json = JSONObject(body)
        val pubkeyB64 = json.getString("x25519_public_key")
        val pubkey = android.util.Base64.decode(pubkeyB64, android.util.Base64.DEFAULT)
        relayPublicKey = pubkey
        pubkey
    }

    private suspend fun getRelayPublicKey(): ByteArray {
        return relayPublicKey ?: fetchInfo()
    }

    /// Register a push subscription for a group topic.
    suspend fun subscribe(
        subscriptionID: String,
        filter: JSONObject,
        pushToken: String,
        notificationKey: ByteArray,
        platform: String
    ) {
        val relayPubkey = getRelayPublicKey()
        val payload = JSONObject().apply {
            put("action", "subscribe")
            put("subscription_id", subscriptionID)
            put("filter", filter)
            put("token", pushToken)
            put("notification_key", android.util.Base64.encodeToString(notificationKey, android.util.Base64.NO_WRAP))
            put("platform", platform)
        }
        sendEncryptedEnvelope(payload, relayPubkey)
    }

    /// Update an existing subscription's filter.
    suspend fun update(subscriptionID: String, filter: JSONObject) {
        val relayPubkey = getRelayPublicKey()
        val payload = JSONObject().apply {
            put("action", "update")
            put("subscription_id", subscriptionID)
            put("filter", filter)
        }
        sendEncryptedEnvelope(payload, relayPubkey)
    }

    /// Unsubscribe.
    suspend fun unsubscribe(subscriptionID: String) {
        val relayPubkey = getRelayPublicKey()
        val payload = JSONObject().apply {
            put("action", "unsubscribe")
            put("subscription_id", subscriptionID)
        }
        sendEncryptedEnvelope(payload, relayPubkey)
    }

    /// Encrypt payload using ephemeral X25519 + AES-256-GCM and POST to /v1/subscription.
    private suspend fun sendEncryptedEnvelope(payload: JSONObject, relayPubkey: ByteArray) = withContext(Dispatchers.IO) {
        val payloadBytes = payload.toString().toByteArray()
        // encryptInvitation uses ephemeral X25519 ECDH — anonymous, no signing
        val envelopeJSON = GroupCrypto.encryptInvitation(
            payload = payloadBytes,
            recipientKeyAgreementPubkey = relayPubkey,
            senderKeyManager = null  // No signing — anonymous registration
        )

        val request = Request.Builder()
            .url("$relayURL/v1/subscription")
            .post(envelopeJSON.toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            val body = response.body?.string() ?: ""
            throw Exception("PN relay error ${response.code}: $body")
        }
    }
}
